src/main/java/seedu/duke/command/ListNoteCommand.java

```diff
@@ -85,61 +87,70 @@ public ListNoteCommand(Boolean isAscendingOrder, ArrayList<String> tags) {
      */
     @Override
     public String execute() {
-        String noteString = "";
+        StringBuilder noteString = new StringBuilder();
         ArrayList<Note> notes = new ArrayList<>();
 
         ArrayList<Note> sortedNotes = (ArrayList<Note>) notebook.getNotes().stream()
-                .filter((s) -> s instanceof Note)
-                .sorted((a, b) -> a.getTitle().toLowerCase().compareTo(b.getTitle().toLowerCase()))
+                .filter(Objects::nonNull)
+                .sorted(Comparator.comparing(a -> a.getTitle().toLowerCase()))
                 .collect(Collectors.toList());
 
         if (tags == null) {
 
             if (isAscendingOrder == null) {
                 for (int i = 0; i < notebook.getNotes().size(); i++) {
-                    noteString += (i + 1) + "." + notebook.getNotes().get(i).toString();
+                    noteString.append(i + 1).append(".")
+                            .append(notebook.getNotes().get(i).getTitle())
+                            .append(InterfaceManager.LS);
                 }
             } else {
-                noteString = getSortedString(noteString, sortedNotes);
-            }
-        } else {
-            Map<Tag, ArrayList<Note>> tag = tagManager.getTagMap();
-
-            // Based on user inputted tags, will store the respective values in an ArrayList
-            // E.g. if user input 2 tags, CS2113 and important, will have 2 ArrayList
-            //      1 for the values corresponding to CS2113 and the other for important tag
-            List<ArrayList<Note>> values = tags.stream()
-                    .map(tag::get)
-                    .collect(Collectors.toList());
-
-            for (int i = 0; i < values.size(); i++) {
-                for (int j = 0; j < values.get(i).size(); j++) {
-                    // Account for duplicates.
-                    // In case an item has both CS2113 and Important tag
-                    if (!notes.contains(values.get(i).get(j))) {
-                        notes.add(values.get(i).get(j));
-                    }
-                }
+                noteString = new StringBuilder(getSortedString(noteString.toString(), sortedNotes));
             }
 
-            // Sort the tagged notes
-            ArrayList<Note> sortedTaggedNotes = (ArrayList<Note>) notes.stream()
-                    .filter((s) -> s instanceof Note)
-                    .sorted((a, b) -> a.getTitle().toLowerCase().compareTo(b.getTitle().toLowerCase()))
-                    .collect(Collectors.toList());
+            return noteString.toString();
+        }
 
+        Map<Tag, ArrayList<Note>> tagMap = tagManager.getTagMap();
+        ArrayList<Tag> tagList = new ArrayList<>();
 
-            if (isAscendingOrder == null) {
-                for (int i = 0; i < notes.size(); i++) {
-                    noteString += (i + 1) + "." + notes.get(i).toString();
+        for (String tag : tags) {
+            tagList.add(tagManager.getTag(tag));
```

Comment:
Have you tested listing notes with a non-existing tag? i.e. list notes with CS2113 tag, but there is no such tag in the map at all? getTag returns null for such case so I think you should only add those that are not null.

```suggestion
            Tag tag = tagManager.getTag(tag);
            if (tag != null) {
                tagList.add(tag);
            }
```